### PR TITLE
Support Asset transfer and opt-in in Runtime

### PR DESCRIPTION
**Pull request by ratik21**
- Added asset opt-in, tests
- Added asset transfer, tests

### CODE REVIEW

File: packages/runtime/src/account.ts [hunk 1]
```diff
@@ -227,6 +235,22 @@ export class StoreAccount implements StoreAccountI {
     }
   }
 
+  // opt-in to asset
+  optInToASA (assetIndex: number, assetHolding: AssetHolding): void {
+    const localAsset = this.assets.get(assetIndex); // fetch asset holding of account
```
Comment (amityadav0) [inline packages/runtime/src/account.ts:240]: how about renaming to `localAsset` to `assetHolding` ?

File: packages/runtime/src/errors/errors-list.ts [hunk 2]
```diff
@@ -308,6 +314,12 @@ const asaErrors = {
     message: `Asset with Index %assetId% not found`,
     title: "Asset Not Found",
     description: "Asset Not Found"
+  },
+  ASSET_HOLDING_NOT_FOUND: {
+    number: 903,
+    message: `Account %address% does not hold asset index %assetId%`,
```
Comment (amityadav0) [inline packages/runtime/src/errors/errors-list.ts:320]:
```suggestion
    message: `Account %address% doesn't hold asset index %assetId%`,
```

File: packages/runtime/src/interpreter/opcode-list.ts [hunk 3]
```diff
@@ -1921,7 +1921,7 @@ export class GetAssetHolding extends Op {
         value = BigInt(assetInfo.amount);
         break;
       case "AssetFrozen":
-        value = stringToBytes(assetInfo["is-frozen"]);
+        value = BigInt(assetInfo["is-frozen"]);
```
Comment (amityadav0) [inline packages/runtime/src/interpreter/opcode-list.ts:1924]: assetInfo["is-frozen"] ?  1n : 0n;
how about using the above ?
i think this is more intuitive.

File: packages/runtime/src/runtime.ts [hunk 4]
```diff
@@ -314,11 +314,35 @@ export class Runtime {
...
+    this.mkAssetCreateTx(name, flags, asset);
     this.store.assetDefs.set(this.assetCounter, sender.addr);
     return this.assetCounter;
   }
 
+  /**
+   * Asset Opt-In for account in Runtime
+   * @param assetIndex Asset Index
+   * @param address Account address to opt-into asset
+   * @param flags txParams
```
Comment (amityadav0) [inline packages/runtime/src/runtime.ts:326]:
```suggestion
   * @param flags Transaction Parameters
```

File: packages/runtime/src/runtime.ts [hunk 5]
```diff
@@ -580,6 +621,29 @@ export class Runtime {
     }
   }
 
+  // transfer ASSET as per transaction parameters
+  transferAsset (txnParam: AssetTransferParam): void {
+    const fromAssetHolding = this.getAssetHolding(txnParam.assetID, txnParam.fromAccount.addr);
+    const toAssetHolding = this.getAssetHolding(txnParam.assetID, txnParam.toAccountAddr);
+
```
Comment (amityadav0) [inline packages/runtime/src/runtime.ts:628]: let's add a check here for asset frozen.

File: packages/runtime/src/account.ts (hunk 1 above)
Comment (ratik21) [reply packages/runtime/src/account.ts:240]: renamed

File: packages/runtime/src/interpreter/opcode-list.ts (hunk 3 above)
Comment (ratik21) [reply packages/runtime/src/interpreter/opcode-list.ts:1924]: I don't have a particular preference. But will update if you think this is more intuitive

File: packages/runtime/src/runtime.ts (hunk 5 above)
Comment (ratik21) [reply packages/runtime/src/runtime.ts:628]: hmm, thanks. So we should also check if asset is frozen or not during opt-in?

Comment (ratik21) [reply packages/runtime/src/runtime.ts:628]: Added test as well.

Comment (amityadav0) [reply packages/runtime/src/runtime.ts:639]: Thanks

File: packages/runtime/src/runtime.ts [hunk 6]
```diff
@@ -580,6 +632,30 @@ export class Runtime {
...
+    this.assertAssetNotFrozen(txnParam.assetID, txnParam.fromAccount.addr);
```
Comment (amityadav0) [inline packages/runtime/src/runtime.ts:640]: does it mean `to` can receive assets even if it is frozen ? could you please confirm ?

Comment (ratik21) [reply packages/runtime/src/runtime.ts:640]: It's probably both. But i can't test it to be 100% sure, as algob doesn't support asset freeze right now. Will validate after we add this support.
```
An Asset Freeze Transaction is issued by the Freeze Address and results in the asset receiver address losing or being granted the ability to send or receive the frozen asset.
```
Adding assert for `to` as well.

File: packages/runtime/src/interpreter/opcode-list.ts (hunk 3 above)
Comment (robert-zaremba) [reply packages/runtime/src/interpreter/opcode-list.ts]: I like @amityadav0  suggestion.

**Review by amityadav0 — APPROVED** [review]

**Pull request merged**